api_controller/models.py

```diff
@@ -1,139 +1,201 @@
+
+
+class ApiController(SingletonModel):
     oauth_credentials = {
         "client_id": settings.STEPIK_CLIENT_ID,
         "client_secret": settings.STEPIK_CLIENT_SECRET,
     }
-    base_url = models.TextField()
-    api_version = 0.1
-    api_host = "https://stepik.org"
+    api_version = models.FloatField(default=0.1)
+    api_host = models.CharField(max_length=255, default="https://stepik.org")
     token = None
 
+    def from_str_to_datetime(self, str):
+        return datetime.strptime(str, "%Y-%m-%dT%H:%M:%SZ")
+
     def stepik_oauth(self):
 
         auth = requests.auth.HTTPBasicAuth(self.oauth_credentials["client_id"], self.oauth_credentials["client_secret"])
         response = requests.post('https://stepik.org/oauth2/token/',
                                  data={'grant_type': 'client_credentials'},
                                  auth=auth)
         self.token = response.json().get('access_token', None)
-        print(self.token)
         if not self.token:
-            print('Unable to authorize with provided credentials')
-            exit(1)
+            return False
+        return True
 
+    # :returns stepik_object or None if any errors occured during call STEPIK API
     def fetch_stepik_object(self, obj_class, obj_id):
         api_url = '{}/api/{}s/{}'.format(self.api_host, obj_class, obj_id)
-        print("API_URL", api_url)
         response = requests.get(api_url,
                                 headers={'Authorization': 'Bearer ' + self.token}).json()
         if '{}s'.format(obj_class) in response:
             return response['{}s'.format(obj_class)][0]
         return None
 
+    def fetch_stepik_objects(self, obj_class, obj_ids, keep_order=True):
+        objs = []
+        load_size = 30  # don't wanna hit HTTP request limits
+        for i in range(0, len(obj_ids), load_size):
+            slice = obj_ids[i:i + load_size]
+            api_url = '{}/api/{}s/?{}'.format(self.api_host, obj_class, "&".join("ids[]={}".format(x) for x in slice))
+            response = requests.get(api_url,
+                                    headers={'Authorization': 'Bearer ' + self.token}).json()
+            objs += response['{}s'.format(obj_class)]
+        if keep_order:
+            return sorted(objs, key=lambda x: obj_ids.index(x['id']))
+        return objs
+
     def fetch_stepik_step(self, pk):
         return self.fetch_stepik_object("step", pk)
 
     def fetch_stepik_lesson(self, pk):
         return self.fetch_stepik_object("lesson", pk)
 
-    def steps_text(self, ids):
+    # :returns: list of tuples(text, update_date)
+    def steps_text_with_dates(self, ids):
         texts = []
-        for id in ids:
-            step = self.fetch_stepik_object("step", id)
-            if step is None or 'text' in step['block']:
+        steps = self.fetch_stepik_objects("step", ids)
+        for step in steps:
+            if step is None or 'text' not in step['block']:
                 texts.append("")
```

Comment:
@KirillMakarov we need something for every step and we don't keep here ids, that's why we need this